plugins/node/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -0,0 +1,749 @@
+              self._config.publishHook!(span, {
+                moduleVersion,
+                exchange,
+                routingKey,
+                content,
+                options,
+                isConfirmChannel: false,
+              }),
+            e => {
+              if (e) {
+                diag.error('amqplib instrumentation: publishHook error', e);
+              }
+            },
+            true
+          );
+        }
+
+        // calling normal channel publish function is only storing the message in queue.
+        // it does not send it and waits for an ack, so the span duration is expected to be very short.
+        const argumentsCopy = [...arguments];
+        argumentsCopy[3] = modifiedOptions;
+        const originalRes = original.apply(this, argumentsCopy as any);
+        span.end();
+        return originalRes;
+      }
+    };
+  }
+
+  private createPublishSpan(
+    self: this,
+    exchange: string,
+    routingKey: string,
+    channel: InstrumentationPublishChannel,
+    options?: amqp.Options.Publish
+  ) {
+    const normalizedExchange = normalizeExchange(exchange);
+
+    const span = self.tracer.startSpan(
+      `${normalizedExchange} -> ${routingKey} send`,
+      {
+        kind: SpanKind.PRODUCER,
+        attributes: {
+          ...channel.connection[CONNECTION_ATTRIBUTES],
+          [SemanticAttributes.MESSAGING_DESTINATION]: exchange,
+          [SemanticAttributes.MESSAGING_DESTINATION_KIND]:
+            MessagingDestinationKindValues.TOPIC,
+          [SemanticAttributes.MESSAGING_RABBITMQ_ROUTING_KEY]: routingKey,
+          [SemanticAttributes.MESSAGING_MESSAGE_ID]: options?.messageId,
+          [SemanticAttributes.MESSAGING_CONVERSATION_ID]:
+            options?.correlationId,
+        },
+      }
+    );
+    const modifiedOptions = options ?? {};
+    modifiedOptions.headers = modifiedOptions.headers ?? {};
+
+    propagation.inject(
+      trace.setSpan(context.active(), span),
+      modifiedOptions.headers
+    );
+
+    return { span, modifiedOptions };
+  }
+
+  private endConsumerSpan(
+    message: InstrumentationMessage,
+    isRejected: boolean | null,
+    operation: EndOperation,
+    requeue: boolean | undefined
+  ) {
+    const storedSpan: Span | undefined = message[MESSAGE_STORED_SPAN];
+    if (!storedSpan) return;
+    if (isRejected !== false) {
+      storedSpan.setStatus({
+        code: SpanStatusCode.ERROR,
+        message:
+          operation !== EndOperation.ChannelClosed &&
+          operation !== EndOperation.ChannelError
+            ? `${operation} called on message${
+                requeue === true
+                  ? ' with requeue'
+                  : requeue === false
+                  ? ' without requeue'
+                  : ''
+              }`
+            : operation,
+      });
+    }
+    this.callConsumeEndHook(storedSpan, message, isRejected, operation);
+    storedSpan.end();
+    delete message[MESSAGE_STORED_SPAN];
```

Comment:
from performance point of view setting it to `undefined` instead deleteing it is better

Reply:
changed to use undefined.
Thanks :)